Firestore/Swift/Source/Codable/DocumentSnapshot+ReadDecodable.swift

```diff
@@ -35,11 +35,9 @@ extension DocumentSnapshot {
   ///     default decoder.
   public func data<T: Decodable>(as type: T.Type,
                                  with serverTimestampBehavior: ServerTimestampBehavior = .none,
-                                 decoder: Firestore.Decoder? = nil) throws -> T? {
+                                 decoder: Firestore.Decoder? = nil) throws -> T {
     let d = decoder ?? Firestore.Decoder()
-    if let data = data(with: serverTimestampBehavior) {
-      return try d.decode(T.self, from: data, in: reference)
-    }
-    return nil
+    let data: Any = data(with: serverTimestampBehavior) ?? NSNull()
```

Comment:
The comment says "By default, server-provided timestamps that have not yet been set to their final value will be returned as `NSNull`. Pass `serverTimestampBehavior` configure this behavior.", but I don't think this is what's happening here.

1. We do pass `serverTeimstampBehaviour` - that much is true
2. If `data(with:)` returns `nil`, we convert this to `NSNull()` - it's not entirely clear to me why.  The following two alternatives compile just the same:
    - `let data: Any = data(with: serverTimestampBehavior) ?? .none`
    -  `let data: Any = data(with: serverTimestampBehavior)`

Reply:
I'll have a look at the comments. The reason for using NSNull() is that the decoder can decode it into the .none case of an Optional. So if the type being decoded is optional, then it will decode just fine, if it is not, it will fail decoding.
Similar trick is used in the current Snapshot decoding for RTDB.

Reply:
Now I've looked a bit more: The initial mention of `NSNull` is from the existing implementation and must be referring to what happens internally when getting data with serverside timestamp values in them.

That comment does not relate to the usage of passing `NSNull` to the unboxing in case there is no data. The latter is what I explained above.

Reply:
I just looked into this particular spot again today. There are two issues I'd like to see resolved:

1) The mention of `NSNull` is a bit confusing. Can we phrase the doc comment differently? Maybe we can just say "The default behaviour is to leave timestamps untouched" or something like this. @schmidt-sebastian can you chime in here to provide guidance?

2) I *think* this code changes the default behaviour in case a document doesn't exist. Previously, when calling `getDocument` for a non-existing document, you would get `nil` as a result. Now, you get a `DecodingError.valueNotFound` instead ("The data couldn’t be read because it is missing." / "Cannot get keyed decoding container -- found null value instead."). To me, these are two different kinds of errors, and this might be misleading for developers. The error message indicates that something with the decoding process, when in fact the document was missing. This can be alleviated by passing in an optional type (`Book?.self`), and having to be more explicit about this might force developers to think more about their expectations when mapping documents, but I'd like to discuss this concern so we can properly document it.